### DISCUSSION TITLE
Stemlight v1.1.4-beta

### DISCUSSION THREAD

**Discussion by ncolyer11** [ncolyer11/Stemlight#18] [General]
### News
- Added and fine-tuned (automatically now) a machine learning algorithm from scratch to optimise dispenser placements for maximum fungus production given a minimum bone meal efficiency:
![image](https://github.com/ncolyer11/Stemlight/assets/90807490/95a4a75d-44ab-43dd-814d-90d41077fa3f)

- All desired features for the playerless core tool have now been implemented.

### Features
- A custom simulated annealing algorithm is now being used to optimise any placement of dispensers on any sized grid of nylium (taking into account a minimum bone meal efficiency) in a **fixed** amount of time, configurable via a new menu toolbar:
![image](https://github.com/ncolyer11/Stemlight/assets/90807490/d82ff955-0123-4764-b248-0f59627ec0db)
  - Configurable parameters for this algorithm, start and end temperature, as well as cooling rate, are automatically calculated using a bit and hint of math inspired from a couple research papers, but primarily this one: https://digitalcommons.calpoly.edu/theses/497/
  - Note, this optimisation doesn't take into account cleared dispensers as that doesn't make practical sense when put into the context of designing a playerless nether tree farm core.

- Rearranged the playerless core GUI to be more horizontally inclined; placing the sliders in a 2x2 grid instead of having them stacked on top of each other, whilst also adding a new output section that displays foliage distribution for an individual, selected (via right-clicking), block:
![image](https://github.com/ncolyer11/Stemlight/assets/90807490/44d62721-ff0c-49f7-9340-25b1d37017c7)
    - Deselect a block by just right-clicking on it again. Note that additional information is displayed for dispensers, so in total that includes: fungi and foliage generated on just the selected block, fungi produced by the selected dispenser, bone meal usage and efficiency, and order and cleared status of the dispenser.
- Middle-click (or ctrl+click for users without a mouse) a dispenser to set it to 'cleared' mode. Such a mode is achieved by placing a piston above a nylium dispenser that fires after every cycle (excluding the last) which has the aim of clearing the block of foliage directly above the nylium dispenser to ensure that it's able to fire almost every cycle (note if it's not the first dispenser to fire for a single cycle, then there's a chance that for each individual cycle, the dispenser could still be blocked):
![image](https://github.com/ncolyer11/Stemlight/assets/90807490/9c866e61-e133-499d-9d04-2fd90906878f)
    - Currently, foliage that's generated but then removed is not counted towards the bone meal savings by retrieving and composting excess foliage. This is planned to be added in a future update.

- Started to add a new help menu to each program that just displays information about the program's purpose and how to use it:
![image](https://github.com/ncolyer11/Stemlight/assets/90807490/0bb3cfac-08c7-4051-9992-d044d7d6d0a4)

### Bug Fixes
- Simulated annealing algorithm only works for the fixed nylium grid size of 5x5 and doesn't take into account minimum bone meal efficiency.
- Optimisation load time is extremely slow for dispenser counts above ~5, causing maximum runtime exceeded errors.
- Increasing max length to 9 on small screens causes the window to go out of frame and makes certain elements of the GUI inaccessible/unviewable.
- Individual block information window shows NAN and/or +/-inf for dispenser foliage counts after selecting a cleared dispenser in close proximity to another dispenser.
- Decreased load time by compiling the application package into multiple files.
- Individual block information would only capture the foliage distribution for the last cycle.

<hr /><em>This discussion was created from the release <a href='https://github.com/ncolyer11/Stemlight/releases/tag/v1.1.4-beta'>Stemlight v1.1.4-beta</a>.</em>

**Comment by SuperLlama88888** [comment]
Yesss now you can put "AI" in the thumbnail and get 10x views!
